3p/iframe-transport-client.js

```diff
@@ -33,61 +33,134 @@ export class IframeTransportClient {
-              this.listener_ &&
-                  this.listener_(event.message, event.transportId);
+              dev().assert(event.creativeId,
+                  'Received malformed event in ' + this.win_.location.href);
+              this.contextFor_(event.creativeId).dispatch(event.message);
             } catch (e) {
               user().error(TAG_,
-                  'Exception in callback passed to onAnalyticsEvent: ' +
-                  e.message);
+                  'Exception in callback passed to onAnalyticsEvent',
+                  e);
             }
           });
         });
   }
 
   /**
+   * Retrieves/creates a context object to pass events pertaining to a
+   * particular creative.
+   * @param {string} creativeId The ID of the creative
+   * @returns {!IframeTransportContext}
+   * @private
+   */
+  contextFor_(creativeId) {
+    return this.creativeIdToContext_[creativeId] ||
+        (this.creativeIdToContext_[creativeId] =
+            new IframeTransportContext(this.win_, this.iframeMessagingClient_,
+            creativeId, this.vendor_));
+  }
+
+  /**
+   * Gets the IframeMessagingClient.
+   * @returns {!IframeMessagingClient}
+   * @VisibleForTesting
+   */
+  getIframeMessagingClient() {
+    return this.iframeMessagingClient_;
+  }
+}
+
+/**
+ * A context object to be passed along with event data.
+ */
+class IframeTransportContext {
+  /**
+   * @param {!Window} win
+   * @param {!IframeMessagingClient} iframeMessagingClient
+   * @param {string} creativeId The ID of the creative that the event
+   *     pertains to.
+   * @param {string} vendor The 3p vendor name
+   */
+  constructor(win, iframeMessagingClient, creativeId, vendor) {
+    /** @private {!IframeMessagingClient} */
+    this.iframeMessagingClient_ = iframeMessagingClient;
+
+    /** @private @const {!Object} */
+    this.baseMessage_ = {creativeId, vendor};
+
+    /** @private {?function(string)} */
+    this.listener_ = null;
+
+    user().assert(win['onNewAmpAnalyticsInstance'] &&
+        typeof win['onNewAmpAnalyticsInstance'] == 'function',
+        'Must implement onNewAmpAnalyticsInstance in ' + win.location.href);
+    win['onNewAmpAnalyticsInstance'](this);
+  }
+
+  /**
    * Registers a callback function to be called when an AMP analytics event
    * is received.
    * Note that calling this a second time will result in the first listener
    * being removed - the events will not be sent to both callbacks.
-   * @param {function(string,string)} callback
+   * @param {!function(string)} listener
    */
-  onAnalyticsEvent(callback) {
-    this.listener_ = callback;
+  onAnalyticsEvent(listener) {
+    this.listener_ = listener;
   }
 
   /**
-   * Gets the IframeMessagingClient
-   * @returns {!IframeMessagingClient}
-   * @VisibleForTesting
+   * Receives an event from IframeTransportClient, and passes it along to
+   * the creative that this context represents.
+   * @param {string} event
+   */
+  dispatch(event) {
```

Comment:
Note: I think we will soon need an event object (IntersectionObserver Entry for example). I am perfectly OK with a string here if we will not introduce breaking change if we migrate to object later.

Reply:
Agreed - definitely do not want to introduce a breaking change in the future.

Reply:
I think there's may a breaking change involved here though. 
If I am a vendor trying to override the `onAnalyticsEvent` method. I would assume all events consist of strings. For example, how can we not break this use case if we change event to an object in future. 
```
onAnalyticsEvent = event => {
  if (event.indexOf('special-prefix')) {
    //do something
  }
}
```

Reply:
My thinking is that we could keep onAnalyticsEvent() for events that are truly about amp-analytics (e.g. come from the "request" part of an amp-analytics tag's config), and IntersectionObserver information could come in via a different way, such as onIntersectionData()